### PR TITLE
Issue #3612: Fixed left curly indentation

### PR DESCRIPTION
**Pull request by soon**
Issue #3612 (related PR #3613)

Diffs: https://soon-checkstyle-diffs.github.io/issue-3612/

~~Unfortunately, I'm unable to create regression diffs on all projects, because checkstyle fails with `OutOfMemory` error while generating report (it fails on master as well).~~

~~I've made reports for guava (https://soon-checkstyle-diffs.github.io/issue-3612/guava/) and jdk (https://soon-checkstyle-diffs.github.io/issue-3612/jdk/) using their checkstyle configs. Is it possible to set different checkstyle configs for different projects at once? Probably, in the properties file?~~

I'm also a little bit confused with the [line number 32 in the test case](https://github.com/checkstyle/checkstyle/pull/3973/files#diff-41124079fe0fabb5d72bbb58a1f17c8cR32). I'm not sure if it should raise a warning.

### CODE REVIEW

**Comment by romani** [comment]
@soon ,

> because checkstyle fails with OutOfMemory error while generating report (it fails on master as well).

sad, please do split generation in few launches(multiple launches of diff tool).

> Is it possible to set different checkstyle configs for different projects at once?

no, just split them in separate launches of this tool.

> Probably, in the properties file?

no way.

**Comment by rnveach** [comment]
> because checkstyle fails with OutOfMemory error while generating report

Indentation is very special. To complete regression I needed to give the JVM almost 4 gigs of memory. This is why we specify `-DMAVEN_OPTS=-Xmx3024m`.
Does it give a stack trace when the `OutOfMemory` happens?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/ArrayInitHandler.java [hunk 1]
```diff
@@ -91,7 +91,7 @@ protected DetailAST getListChild() {
     protected IndentLevel getChildrenExpectedIndent() {
         final IndentLevel expectedIndent =
             new IndentLevel(getIndent(), getIndentCheck().getArrayInitIndent(),
-                    getIndentCheck().getLineWrappingIndentation());
+                getLineWrappingIndentation());
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/ArrayInitHandler.java]: I assume this is just a code simplification change and not necessary to this issue?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/BlockParentHandler.java [hunk 2]
```diff
@@ -151,13 +151,26 @@ protected DetailAST getRightCurly() {
      * Check the indentation of the left curly brace.
      */
     protected void checkLeftCurly() {
-        // the lcurly can either be at the correct indentation, or nested
-        // with a previous expression
-        final DetailAST lcurly = getLeftCurly();
-        final int lcurlyPos = expandedTabsColumnNo(lcurly);
+        checkCurly("lcurly", getLeftCurly());
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/BlockParentHandler.java]: I assume everything in this file is the same, just a code simplification change?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/ArrayInitHandler.java (hunk 1 above)
Comment (soon) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/ArrayInitHandler.java]: Sure

Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/ArrayInitHandler.java]: Having issue fix and big code simplification changes in 1 commit may make this too hard to review.
Unless @romani wants otherwise, I think we should split this into 2 commits.
1st commit is minor with all code simplifications.
2nd commit is issue with all code related to issue fix.

Comment (soon) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/ArrayInitHandler.java]: @rnveach No problem. Should I move simplifications to another PR?

Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/ArrayInitHandler.java]: @soon it might be better just we can verify issue change is the only one affecting violation changes.

**Comment by soon** [comment]
@rnveach Excluded all redundant code, please, review

**Comment by rnveach** [comment]
@soon Can you try to run regression on as many projects as possible?
Indentation is very complex check and you never know where you will find the one project with something interesting in it.

**Comment by soon** [comment]
@rnveach Yes, I'm just running it. Thank you for the `MAVEN_OPTS`, I tried to specify java and groovy opts and totally forgot about mvn.

**Comment by soon** [comment]
@rnveach Sorry for the long reply. I've updated diffs link (https://soon-checkstyle-diffs.github.io/issue-3612/)

1. findbugs (https://soon-checkstyle-diffs.github.io/issue-3612/findbugs/xref/home/soon/Src/Java/checkstyle-contribution/checkstyle-tester/repositories/findbugs/bluej/src/edu/umd/cs/findbugs/bluej/ResultsFrame.java.html#L97) - looks strange, again `forceStrictCondition`?
2. checkstyle (https://soon-checkstyle-diffs.github.io/issue-3612/checkstyle-with-excludes/xref/home/soon/Src/Java/checkstyle-contribution/checkstyle-tester/repositories/checkstyle-with-excludes/src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/InputInvalidClassDefIndent.java.html#L133), should we update comment with `//indent:8 exp:10,14,18 warn`?

All other diffs looks fine for me. Could you, please, review them?

**Comment by soon** [comment]
@rnveach, @romani, should I add something else?

**Comment by romani** [comment]
> looks strange, again forceStrictCondition?

http://checkstyle.sourceforge.net/config_misc.html#Indentation `forceStrictCondition` is false by default, in your config it is not define - false.
Previous violation was valid to  some extend ...... @rnveach , do we consider blocks of anymous classes as line wrapping or as child block. I think we are treated anonymous body as wrapping.
if that is true, so this case is false-negativem and violation need to be restored. 

> should we update comment with //indent:8 exp:10,14,18 warn?

I see no problem with update.

**Comment by rnveach** [comment]
> do we consider blocks of anymous classes as line wrapping or as child block

It is probably both. We do alot of double validation in some cases, this is why we added the field `incorrectIndentationLines` to prevent duplicate violations on same line.

If `new` starts the line it is validated as part of the block in `MethodCallHandler.checkExpressionSubtree`.
We then call it again in `MethodCallHandler.checkWrappingIndentation` for wrapping multiple lines.

**Comment by rnveach** [comment]
@romani Sorry for yesterday.
https://soon-checkstyle-diffs.github.io/issue-3612/checkstyle/index.html#A1

10 is correct, as we allow the same indentation as the previous line for the brace. I can see people wanting this behavior.
I believe 18 is correct if we believe `{` is a wrap of the method declaration, which I think it should be. It should be no different than if the parameters go on the next line..

The question then is 14 correct. It is probably just a gray area from lack of documentation on `basicOffset` vs `lineWrappingIndentation`. If `basicOffset` should only be used on brace children indentation, then it is wrong.
Violation for is coming from `MethodDefHandler.checkChildren` which is checking the children of line 118. This seems weird to me that it would get a violation so far in the statement.
We get a second violation from `MethodCallHandler.checkExpressionSubtree` but we are hiding it because of the 1 violation per line rule. It just wants `8, 14` [because it specifically uses `basicOffset`](https://github.com/checkstyle/checkstyle/blob/6e8d1e06728b7defbd734253b3818a33676b62bc/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/MethodCallHandler.java#L184).

In my opinon, we should define right here and now what is `basicOffset` and what is `lineWrappingIndentation` and [specify what it is in our other issue](https://github.com/checkstyle/checkstyle/issues/3160).
As for `checkChildren`, I question why it is printing violations on lines that are not the start of the statements. I also question, why we need it since we are duplicating verifying some lines. I think its purpose is for children we aren't validating otherwise but I we would have to look into this more.

**Comment by romani** [comment]
@rnveach , please comment a https://soon-checkstyle-diffs.github.io/issue-3612/openjdk8/index.html#A3 that I referenced in my post.

> In my opinon, we should define right here and now what is basicOffset and what is lineWrappingIndentation

http://checkstyle.sourceforge.net/config_misc.html#Indentation
`basicOffset` is offset of statements on certain level. Level is created braces.
`lineWrappingIndentation` is continuation of statement on new line, it is minimal indentation we require, use can indent even further to the right if he wants.

**Comment by rnveach** [comment]
> please comment a https://soon-checkstyle-diffs.github.io/issue-3612/openjdk8/index.html#A3 that I referenced in my post.
> lineWrappingIndentation is continuation of statement on new line

Example is nearly the same.
`{` is part of the statement, just like a method definition. Only the fields/methods inside the class block are considered children, imo. It is mostly identical to lambdas. `->` is not a child, just part of the statement.
`lineWrappingIndentation` is set to 8, so it is indented correctly. There should be no violation.
